nerdery-ecommerce/src/products/products.service.ts

```diff
@@ -115,5 +135,71 @@ export class ProductsService {
     });
   }
 
+  async create(input: CreateProductInput) {
+    const { categoryId, ...rest } = input;
+
+    if (!categoryId) {
+      throw new BadRequestException('Category is required');
+    }
+
+    if (!await this.categoriesService.doesCategoryExist(categoryId)) {
+      throw new BadRequestException('Category does not exist');
+    }
+
+
+    return await this.prisma.product.create({
+      data: {
+        ...rest,
+        category: {
+          connect: {
+            id: categoryId
+          }
+        },
+      },
+      include: { productVariations: true, category: true },
+    });
+  }
+
+  async update(input: UpdateProductInput) {
+    const { categoryId, ...rest } = input;
+
+    if (categoryId) {
+      if (!this.categoriesService.doesCategoryExist(categoryId)) {
+        throw new Error('Category does not exist');
+      }
+    }
+
+    await this.productsHelperService.findProductByIdAndValidate({ id: input.id });
+
+    return this.prisma.product.update({
+      where: { id: input.id },
+      data: {
+        ...rest,
+        category: {
+          connect: {
+            id: categoryId
+          }
+        },
+      },
+      include: { productVariations: true, category: true },
+    });
+  }
+
+  async delete(id: string) {
+    await this.productsHelperService.findProductByIdAndValidate({ id }, false, false);
+
+    await this.prisma.productVariation.updateMany({ where: { productId: id }, data: { isDeleted: true } });
+    return await this.prisma.product.update({ where: { id }, data: { isDeleted: true } });
+  }
+
+  async toggleIsEnabled(id: string, isEnabled: boolean) {
+    await this.productsHelperService.findProductByIdAndValidate({ id }, false, false);
+
+    await this.prisma.productVariation.updateMany({ where: { productId: id }, data: { isEnabled: isEnabled } });
+    if (isEnabled) {
+      await this.productsHelperService.recalculateProductMinMaxPrices([id]);
```

Comment:
Does this method cover the case when no ProductVariations in a Product?

```ts
    await this.productsHelperService.recalculateProductMinMaxPrices([prodVariation.productId]);

```

If that's the case, I need to set maxPrice and minPrice to zero.
This way I make sure the **product query does not retrieve this product even if the product is enable**

Reply:
Solved!

```ts
 async recalculateProductMinMaxPrices(productIds: string[]) {

        const products = await this.prisma.product.findMany({
            where: {
                id: {
                    in: productIds,
                },
                isDeleted: false,
                isEnabled: true,
            },
            include: { productVariations: true },
        });

        const updatePromises = products.map(async (product) => {

            const filteredProductVariations = product.productVariations.filter(p => p.isEnabled && !p.isDeleted);
            const prices = filteredProductVariations.map(p => Number(p.price));

            const minPrice = prices.length > 0 ? Math.min(...prices) : 0;
            const maxPrice = prices.length > 0 ? Math.max(...prices) : 0;

            //If no valid product variations, prices are 0. So users can't add to cart, nor see the product on "products" query
            let calculatedData = { minPrice, maxPrice };
            if (minPrice) calculatedData['minPrice'] = minPrice;
            if (maxPrice) calculatedData['maxPrice'] = maxPrice;

            await this.prisma.product.update({
                where: { id: product.id },
                data: calculatedData,
            });
        });

        await Promise.allSettled(updatePromises);
    }```